### ISSUE TITLE
Request: Support Svelte JS labels

### ISSUE THREAD

**Issue by n-smits** [jashkenas/coffeescript#5367]
There was no need for JS labels before, but Svelte found a good use for the syntax. Currently it can only be passed with backticks, but it's very limited and unwieldy.
```coffee
`$: {`
foo 'hi' # must not properly indent
`}`
```
All other fail
```coffee
`$:` foo 'hi'
# produces ({ $: foo 'hi' })

`$:`
foo 'hi'
# erroneous
# $:;
# foo('hi');

`$: {` foo 'hi' `}`
# won't compile
```
Simplest and very Coffee solution would be to spell out the syntax name, but honestly whatevs.

```coffee
label $: foo 'hi' # or = or space
label $
  foo 'hi'
  foo 'hi'
```

It's true this isn't idiomatic use and that it's a solitary kinda-framework example. But, it is actually using the standard syntax, and its unidiomatic use became innovative, because unlike before, it's actually a good use. Past addition of JSX could be considered similar, and a good place for it in documentation (since it is after all a rare good example of Js labels, so it's mostly about Svelte).

In my very humble general opinion here, Coffeescript should adapt here and there because Javascript is a hackable text based language for better and for worse. Otherwise is excessively conservative and necessarily keeps the bad part of Js (Js is what it is), while losing the better part (it's hackable, here for the better). Good guideline for when is enough of offshoots should include standard syntax when it's a good use even if an unconventional one. (So-so unconvencional actually, it's not like Svelte is so unheard of.)

In the meantime, I'm going to check out [Sweet.js `syntax`](https://www.sweetjs.org/doc/tutorial.html#sweet-let).

**Comment by edemaine** [comment]
I agree that Svelte is an important growing use-case to support. Thanks for raising the issue.

Unfortunately, `label $: foo 'hi'` already has a meaning in CoffeeScript, namely, `label({"$": foo('hi')})`.

The following doesn't have a meaning yet though:

```coffee
label $
  foo 'hi'
```

This matches the syntax style of `if` and `for` and so on, which seems good. This suggests that the inline syntax could be:

```coffee
label $ then foo 'hi'
```

which currently has no meaning, so that seems good.

**Comment by GeoffreyBooth** [comment]
One of the primary reasons for adding support for new features is compatibility. Users shouldn’t need to have to choose between CoffeeScript and Svelte, for example; we added support for JSX (and before that, ES6 classes) to enable full compatibility with React. So yes, we should find some way to support this.

I don’t know about this particular proposed syntax, though. Since `label $` compiles, it feels awfully brittle for users to know they need a second line (or `then`) to get the desired JS label syntax rather than `label($)`. I think it would be better if we could find some new symbol that currently doesn’t compile at all, or always throws at the start of a line; but not `:=`, as that one might get used either for static types or for block assignment, if we ever add either of those.

**Comment by johndeighan** [comment]
Very happy that this is even being discussed. I'm working on an extension to sveltekit and my way around this (which uses backticks) is the most fragile part of my code.

**Comment by n-smits** [comment]
Why the extra colon? As for `:label` (with or without extra colon), I still like it a week later

**Comment by johndeighan** [comment]
The first thing you should be asking is whether your intent is to support the concept of labels in general, or if what you really want to support is reactive assignments and/or statements. Most older uses for labels were very bad (think 'goto'). However, at the end of this post, I'll show a use case that I used often in Perl.

If it's reactive assignments that you want to support, then here's a syntax that I use in something I'm working on. I have no idea if it would work in CoffeeScript - just something to think about:

`doubled <== 2 * count`

This means "whenever a variable in the expression following <== changes, recompute the variable 'doubled'

```
<==
	ctx.fillStyle = $prefs.color
	ctx.fillRect(10, 10, 80, 80)

```
This means whenever any variable in the enclosed code changes, re-run the code.

Personally, I rarely use the 2nd syntax since it's possible to just call a function using the 1st as long as you pass any variables that you want to trigger a re-run to that function. But then you have to define the function somewhere, so a bit less convenient.

It seems like supporting this syntax, assuming it's converted directly to $: <statement> or $: { code block} in the JS would only be supporting svelte programmers. But I think it's also possible to convert it to plain JS code that accomplishes the same thing. A bit complex, I know, but ultimately, svelte does just that.

Now, the argument for simply supporting labels, which would be simpler to implement anyway. Here is a pattern that I've often used in Perl (this may not be valid Perl, but the intent is obvious):

```
FILE:
for filename in lFiles
   content = readFile(filename)
   lLines = content.split('\n')
   LINE:
   for line in lLines
      next LINE if (line == '__END__')
      procLine(line)

```
Now, as written, it will skip any lines like '__END__'. But if instead, I wanted to skip the rest of the current file, I would just change 'next LINE' to 'next FILE'. The point is that when I have nested loops, being able to put a label on each loop allows me to specify unambiguously which loop iteration I'm prematurely exiting. I think this is basically the only good use case I've found for labels. Well, except for supporting svelte, where it's used less as a true label, and more as a way of tagging code as being 'reactive'.

**Comment by gcxfd** [comment]
any update ? 

why not use  $~ for $:

**Comment by gcxfd** [comment]
<img width="1407" alt="image" src="https://user-images.githubusercontent.com/69608605/160052710-e4700253-16dd-4776-993b-ae41a35855e7.png">
https://livescript.net/#functions
I see livescript support label , but It seems dead ...

**Comment by gcxfd** [comment]
I write a coffeescript label patch for [svelte](https://svelte.dev/)

https://github.com/rmw-lib/coffee-label-patch

Make coffeescript support :label syntax, so it can be used for [svelte](https://svelte.dev/)[](https://github.com/rmw-lib/coffee-label-patch#install).

It's too hard to modify the compiler based on lex, ast, and I did this based on string substitution.

Just a quick note, this is a crude hack, but it does work.

I look forward to coffee officially adding this syntax.

## Use

```coffee
#!/usr/bin/env coffee

import CoffeeScript from "coffeescript"
import hack from '@rmw/coffee-label-patch'
hack CoffeeScript

console.log CoffeeScript.compile(
  """
do =>
  y = 0

  :$ x=y*2

  :$ if y>2 then x+=y else x-=y

  :$
    if x > y
      x = y/2
    else
      x = y+9
    x += 1

  :out
    for i in [1,2,3]
      for j in [4,5,6]
        console.log i,j
        if i > 1
          break out
  return
  """
  bare:true
)
```

output :

```js
(() => {
  var i, j, k, l, len, len1, ref, ref1, x, y;
  y = 0;
  $ : x = y * 2;
  $ : y > 2 ? x += y : x -= y;
  $ : {
    if (x > y) {
      x = y / 2;
    } else {
      x = y + 9;
    }
    x += 1;
  }
  out : {
    ref = [1, 2, 3];
    for (k = 0, len = ref.length; k < len; k++) {
      i = ref[k];
      ref1 = [4, 5, 6];
      for (l = 0, len1 = ref1.length; l < len1; l++) {
        j = ref1[l];
        console.log(i, j);
        if (i > 1) {
          break out;
        }
      }
    }
  }
})();
```

**Comment by adminy** [comment]
svelte 5 removes the need for the label hack with runes.

**Comment by SuperPauly** [comment]
Hi all,

I've literally just discovered Coffee Script about 30 minutes ago so I apologize for the noob question but I'm curious to the current compatibility of Svelte and Coffee Script at the moment?

Dose Svelte 5 work with Coffee Script? If they do work together what's the file extension? .svelte or .coffee?

Thanks.

**Comment by adminy** [comment]
@SuperPauly It won't work as both `.svelte` and `.coffee` are transpiling typescript/coffeescript to another form of javascript. If you pipe the output of whatever compiled .coffee looks like into .svelte maybe it could work, but they'd probably throw undefined $effect or $state errors at least in coffeescript, unless there was a way to silence them. Also you'd need a vite/webpack transformer that can do all that if you want it to work like a framework with all the bells and whistles.

**Comment by vendethiel** [comment]
> but they'd probably throw undefined $effect or $state errors at least in coffeescript

That shouldn't be an issue.
